File: src/lib/prewarm/next-episode-file.ts
/**
 * Which file inside an already-held torrent IS a given episode.
 *
 * The expensive part of "play the next episode" is deciding it needs to be
 * fetched at all. When the episode the viewer is about to watch is already
 * sitting inside the season pack currently on screen, the honest answer is
 * "same torrent, this file" — no indexer search, no acquisition, no new swarm.
 *
 * Nothing here parses episodes on its own: it reuses {@link packEpisodeFiles},
 * which already knows how to ignore featurettes, samples and non-video junk,
 * and the persisted verified-file list every other disk-truth reader uses.
 *
 * Pure: no filesystem, no database.
 */
import path from "node:path";
import { heldFilesFromVerifiedJson } from "@/lib/library/deletion-plan";
import { packEpisodeFiles } from "@/lib/torrents/pack-episode-files";

/** The columns this needs off an `EngineTorrent` row. */
export type EpisodeFileRow = {
  savePath: string | null;
  verifiedFilesJson: string | null;
};

/**
 * The torrent-relative path of an absolute on-disk file.
 *
 * `verifiedFilesJson` records absolute paths; every playback surface (the
 * stream manifest, `/api/stream`, `/api/playback/plan`) addresses files by the
 * path *inside* the torrent. Returns null when the file is not contained by the
 * save root — never a guess, never an absolute path a route would reject.
 */
export function torrentRelativeFilePath(
  savePath: string | null | undefined,
  absolutePath: string,
): string | null {
  const root = savePath?.trim();
  if (!root || !path.isAbsolute(absolutePath)) return null;
  const relative = path.relative(path.resolve(root), path.resolve(absolutePath));
  if (!relative || relative.startsWith("..") || path.isAbsolute(relative)) {
    return null;
  }
  return relative.split(path.sep).join("/");
}

/**
 * The torrent-relative path of `season`×`episode` inside this torrent, or null.
 *
 * Null means "no deterministic answer here" — a torrent that is not a pack for
 * that season, an episode whose file has not been verified yet, or a file that
 * lives outside the recorded save root.
 */
export function episodeFileInTorrent(
  row: EpisodeFileRow,
  season: number | null | undefined,
  episode: number | null | undefined,
): string | null {
  if (typeof season !== "number" || typeof episode !== "number") return null;
  const files = heldFilesFromVerifiedJson(row.verifiedFilesJson).map((file) => ({
    path: file.path,
    size: file.sizeBytes,
  }));
  if (files.length === 0) return null;
  const absolute = packEpisodeFiles(files, season).get(episode);
  if (!absolute) return null;
  return torrentRelativeFilePath(row.savePath, absolute);
}

File: server/TorrentFlow.Media/Features/Prewarm/NextEpisodeFile.cs
using System.Text.Json;

namespace TorrentFlow.Media.Features.Prewarm;

/// <summary>Which file inside an already-held torrent is a given episode (src/lib/prewarm/next-episode-file.ts). Pure.</summary>
public static class NextEpisodeFile
{
    /// <summary>
    /// Files recorded in VerifiedFilesJson. Accepts the .NET engine's manifest (<c>{path, size, mtimeMs, fullPath}</c>,
    /// fullPath absolute) and the Next.js shape (<c>{path, size}</c>, path absolute).
    /// </summary>
    public static List<(string Path, long Size)> HeldFiles(string? verifiedFilesJson)
    {
        List<(string, long)> output = [];
        if (string.IsNullOrWhiteSpace(verifiedFilesJson)) return output;
        try
        {
            using var doc = JsonDocument.Parse(verifiedFilesJson);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return output;
            foreach (var entry in doc.RootElement.EnumerateArray())
            {
                if (entry.ValueKind != JsonValueKind.Object) continue;
                string? path = null;
                if (entry.TryGetProperty("fullPath", out var full) && full.ValueKind == JsonValueKind.String) path = full.GetString();
                if (string.IsNullOrEmpty(path) && entry.TryGetProperty("path", out var p) && p.ValueKind == JsonValueKind.String) path = p.GetString();
                if (string.IsNullOrEmpty(path)) continue;
                long size = 0;
                if (entry.TryGetProperty("size", out var s) && s.ValueKind == JsonValueKind.Number && s.TryGetInt64(out var n) && n >= 0) size = n;
                output.Add((path, size));
            }
        }
        catch (JsonException)
        {
            output.Clear();
        }
        return output;
    }

    /// <summary>The torrent-relative (forward-slash) path of an absolute file, or null when it is outside the save root.</summary>
    public static string? TorrentRelativeFilePath(string? savePath, string absolutePath)
    {
        var root = savePath?.Trim();
        if (string.IsNullOrEmpty(root) || !Path.IsPathFullyQualified(absolutePath)) return null;
        var relative = Path.GetRelativePath(Path.GetFullPath(root), Path.GetFullPath(absolutePath));
        if (relative.Length == 0 || relative == "." || relative.StartsWith("..", StringComparison.Ordinal) || Path.IsPathRooted(relative)) return null;
        return relative.Replace('\\', '/');
    }

    /// <summary>The torrent-relative path of season×episode inside this torrent, or null (no deterministic answer).</summary>
    public static string? EpisodeFileInTorrent(string? savePath, string? verifiedFilesJson, int? season, int? episode)
    {
        if (season is not { } s || episode is not { } e) return null;
        var files = HeldFiles(verifiedFilesJson);
        if (files.Count == 0) return null;
        if (!ReleaseText.PackEpisodeFiles(files, s).TryGetValue(e, out var absolute)) return null;
        return TorrentRelativeFilePath(savePath, absolute);
    }
}
